### PR TITLE
Build SCIM2 endpoint URLs and resolve tenant domain for tenant context mode

### PR DESCRIPTION
**Pull request by Abilashini**
Related to https://github.com/wso2/product-is/issues/8404

### CODE REVIEW

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentitySCIMManager.java [hunk 1]
```diff
@@ -66,6 +65,9 @@ private IdentitySCIMManager() throws CharonException {
      */
     public static IdentitySCIMManager getInstance() throws CharonException {
 
+        if (IdentityTenantUtil.isTenantQualifiedUrlsEnabled()) {
+            return new IdentitySCIMManager();
+        }
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentitySCIMManager.java:70]: Any reason to change this method?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java [hunk 2]
```diff
@@ -66,39 +70,54 @@
...
+            }
         } else {
-            scimURL = IdentityUtil.getServerURL(SCIMCommonConstants.SCIM2_ENDPOINT, true, true);
+            String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+            if (isNotASuperTenantFlow(tenantDomain)) {
+                scimURL = IdentityUtil.getServerURL(SCIMCommonConstants.TENANT_URL_SEPERATOR + tenantDomain +
+                        SCIMCommonConstants.SCIM2_ENDPOINT, true, true);
+            } else {
+                scimURL = IdentityUtil.getServerURL(SCIMCommonConstants.SCIM2_ENDPOINT, true, true);
+            }
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java:120]: Do we need this part? Inside the ServiceURLBuilder also it's checking the `isTenantQualifiedUrlsEnabled()` config. So if there is no additional config file used we don't need to specifically check the `isTenantQualifiedUrlsEnabled()` again here. If so we can get rid of the old `IdentityUtil.getServerURL()` completely.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java [hunk 3]
```diff
@@ -107,16 +126,38 @@ private static boolean isNotASuperTenantFlow(String tenantDomain) {
...
+            try {
+                scimURL = ServiceURLBuilder.create().addPath(SCIMCommonConstants.SCIM2_ENDPOINT).build()
+                        .getAbsolutePublicURL();
+            } catch (URLBuilderException e) {
+                throw new IdentitySCIMException("Error occurred while building the SCIM2 endpoint with tenant " +
+                        "qualified URL.", e);
+            }
+        } else {
+            scimURL = IdentityUtil.getServerURL(SCIMCommonConstants.SCIM2_ENDPOINT, true, true);
+        }
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java:142]: Same as the above comment.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java [hunk 4]
```diff
@@ -107,16 +126,38 @@ private static boolean isNotASuperTenantFlow(String tenantDomain) {
...
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java:159]: Same as the above comment.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentitySCIMManager.java (hunk 1 above)
Comment (Abilashini) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentitySCIMManager.java:70]: IdentitySCIMManager object is instantiated during server startup and set the endpoint URLs in AbstractResourceManager with the super tenant URLs at [1]. AbstractResourceManager is accessed by charon to get the endpoint URLs. So for tenants as well the endpoint URLs will the super tenant URLs. 
[1] https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/263/files/42a6b93ef0c49de05581e2fe2dc57dd4a3820aea#diff-609fb5ae29747e7d7564f70cbaa60e44R105

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java (hunk 2 above)
Comment (Abilashini) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java:120]: If the configuration is disabled ServiceURLBuilder doesn't return the tenant domain

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentitySCIMManager.java (hunk 1 above)
Comment (emswbandara) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentitySCIMManager.java:70]: this will instantiate a new instance of IdentitySCIMManager, everytime getInstance() is called. Better to have a instance variable and instantiate only if that variable is null.

File: components/org.wso2.carbon.identity.scim2.provider/src/main/java/org/wso2/carbon/identity/scim2/provider/resources/BulkResource.java [hunk 5]
```diff
@@ -66,7 +66,7 @@ public Response createUser(@HeaderParam(SCIMProviderConstants.AUTHORIZATION) Str
             encoder = identitySCIMManager.getEncoder();
 
             // obtain the user store manager
-            UserManager userManager = IdentitySCIMManager.getInstance().getUserManager();
+            UserManager userManager = identitySCIMManager.getUserManager();
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.provider/src/main/java/org/wso2/carbon/identity/scim2/provider/resources/BulkResource.java:69]: why was this change required?

Comment (Abilashini) [reply components/org.wso2.carbon.identity.scim2.provider/src/main/java/org/wso2/carbon/identity/scim2/provider/resources/BulkResource.java:69]: To avoid instantiating IdentitySCIMManager twice in the same method

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentitySCIMManager.java (hunk 1 above)
Comment (Abilashini) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentitySCIMManager.java:70]: We should instantiate IdentitySCIMManager every time if we want to have tenant qualified URLs. 
Anyways we will be doing this only if tenant context is enabled. otherwise, it will be instantiated only once and the same instance will be used for the follow-up calls.

Comment (malithie) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentitySCIMManager.java:70]: We should not re-instantiate. Rather we should fix getters or add getters if not available to build url based on context

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java (hunk 2 above)
Comment (malithie) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java:120]: So in the legacy way, we still had the tenant qualified complete url built ?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java [hunk 6]
```diff
@@ -177,12 +218,22 @@ public static void setThreadLocalIsManagedThroughSCIMEP(Boolean value) {
     }
 
     public static String getGlobalConsumerId() {
+
+        if (IdentityTenantUtil.isTenantQualifiedUrlsEnabled()) {
+            return IdentityTenantUtil.getTenantDomainFromContext();
```
Comment (malithie) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java:223]: Wondering whether it's good to return tenant domain from privileged carbon context if thread local is not available

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java (hunk 2 above)
Comment (Abilashini) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonUtils.java:120]: yes

**Comment by Abilashini** [comment]
Re-initiated and fixed at https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/265
